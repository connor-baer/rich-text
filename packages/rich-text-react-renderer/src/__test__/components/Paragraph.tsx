import React, { FunctionComponent, ReactNode } from 'react';

type ParagraphProps = {
  children: ReactNode;
};

const Paragraph: FunctionComponent<ParagraphProps> = ({ children }) => {
  return <p style={{ color: 'red' }}>{children}</p>;
};

export default Paragraph;
