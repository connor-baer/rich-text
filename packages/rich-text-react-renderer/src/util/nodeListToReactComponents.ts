import { ReactNode } from 'react';
import { helpers, Mark } from '@contentful/rich-text-types';
import { CommonNode, Options } from '..';
import { appendKeyToValidElement } from './appendKeyToValidElement';

export function nodeListToReactComponents(nodes: CommonNode[], options: Options): ReactNode {
  return nodes.map(
    (node: CommonNode, index: number): ReactNode => {
      return appendKeyToValidElement(nodeToReactComponent(node, options), index);
    },
  );
}

export function nodeToReactComponent(node: CommonNode, options: Options): ReactNode {
  const { renderNode, renderMark } = options;
  if (helpers.isText(node)) {
    return node.marks.reduce((value: ReactNode, mark: Mark): ReactNode => {
      if (!renderMark[mark.type]) {
        return value;
      }
      return renderMark[mark.type](value);
    }, node.value);
  } else {
    if (!node.nodeType || !renderNode[node.nodeType]) {
      return null;
    }
    return renderNode[node.nodeType](node, nodeListToReactComponents(node.content, options));
  }
}
