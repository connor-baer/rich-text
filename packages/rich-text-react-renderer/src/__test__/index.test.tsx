import React from 'react';
import { BLOCKS, Document, INLINES, MARKS } from '@contentful/rich-text-types';

import { documentToReactComponents, Options } from '..';

import {
  embeddedEntryDoc,
  headingDoc,
  hrDoc,
  hyperlinkDoc,
  inlineEntityDoc,
  invalidMarksDoc,
  invalidTypeDoc,
  marksDoc,
  multiMarkDoc,
  olDoc,
  paragraphDoc,
  quoteDoc,
  ulDoc,
} from './documents';
import Paragraph from './components/Paragraph';
import Strong from './components/Strong';
import { appendKeyToValidElement } from '../util/appendKeyToValidElement';

describe('documentToReactComponents', () => {
  it('returns an empty array when given an empty document', () => {
    const document: Document = {
      nodeType: BLOCKS.DOCUMENT,
      data: {},
      content: [],
    };

    expect(documentToReactComponents(document)).toEqual([]);
  });

  it('renders nodes with default node renderer', () => {
    const docs: Document[] = [
      paragraphDoc,
      headingDoc(BLOCKS.HEADING_1),
      headingDoc(BLOCKS.HEADING_2),
    ];

    docs.forEach(doc => {
      expect(documentToReactComponents(doc)).toMatchSnapshot();
    });
  });

  it('renders marks with default mark renderer', () => {
    const docs: Document[] = [
      marksDoc(MARKS.ITALIC),
      marksDoc(MARKS.BOLD),
      marksDoc(MARKS.UNDERLINE),
      marksDoc(MARKS.CODE),
    ];

    docs.forEach(doc => {
      expect(documentToReactComponents(doc)).toMatchSnapshot();
    });
  });

  it('renders multiple marks with default mark renderer', () => {
    const doc: Document = multiMarkDoc();
    expect(documentToReactComponents(doc)).toMatchSnapshot();
  });

  it('renders nodes with passed custom node renderer', () => {
    const options: Options = {
      renderNode: {
        [BLOCKS.PARAGRAPH]: (node, children) => <Paragraph>{children}</Paragraph>,
      },
    };
    const document: Document = paragraphDoc;
    expect(documentToReactComponents(document, options)).toMatchSnapshot();
  });

  it('renders marks with the passed custom mark rendered', () => {
    const options: Options = {
      renderMark: {
        [MARKS.BOLD]: text => <Strong>{text}</Strong>,
      },
    };
    const document: Document = marksDoc(MARKS.BOLD);

    expect(documentToReactComponents(document, options)).toMatchSnapshot();
  });

  it('does not render unrecognized marks', () => {
    const document: Document = invalidMarksDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders empty node if type is not recognized', () => {
    const document: Document = invalidTypeDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders default entry link block', () => {
    const entrySys = {
      sys: {
        id: '9mpxT4zsRi6Iwukey8KeM',
        link: 'Link',
        linkType: 'Entry',
      },
    };
    const document: Document = embeddedEntryDoc(entrySys);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders ordered lists', () => {
    const document: Document = olDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders unordered lists', () => {
    const document: Document = ulDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders blockquotes', () => {
    const document: Document = quoteDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders horizontal rule', () => {
    const document: Document = hrDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('does not crash with inline elements (e.g. hyperlink)', () => {
    const document: Document = hyperlinkDoc;

    expect(documentToReactComponents(document)).toBeTruthy();
  });

  it('renders hyperlink', () => {
    const document: Document = hyperlinkDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders asset hyperlink', () => {
    const asset = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Asset',
        },
      },
    };
    const document: Document = inlineEntityDoc(asset, INLINES.ASSET_HYPERLINK);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
  it('renders entry hyperlink', () => {
    const entry = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Entry',
        },
      },
    };
    const document: Document = inlineEntityDoc(entry, INLINES.ENTRY_HYPERLINK);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
  it('renders embedded entry', () => {
    const entry = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Entry',
        },
      },
    };
    const document: Document = inlineEntityDoc(entry, INLINES.EMBEDDED_ENTRY);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
});

describe('appendKeyToValidElement', () => {
  it('appends keys to default React components', () => {
    expect(appendKeyToValidElement(<div />, 0)).toHaveProperty('key', '0');
  });

  it('appends keys to custom React components', () => {
    expect(appendKeyToValidElement(<Paragraph>hello world</Paragraph>, 0)).toHaveProperty(
      'key',
      '0',
    );
  });

  it('does not overwrite user specified keys', () => {
    expect(appendKeyToValidElement(<div key={'xyz'} />, 0)).toHaveProperty('key', 'xyz');
  });

  it('does not add keys to text nodes', () => {
    expect(appendKeyToValidElement('hello world', 0)).not.toHaveProperty('key');
  });

  it('does not add keys to node arrays', () => {
    expect(appendKeyToValidElement([<div key={0} />, <div key={1} />], 0)).not.toHaveProperty(
      'key',
    );
  });

  it('does not add keys to null', () => {
    expect(appendKeyToValidElement(null, 0)).toBeNull();
  });
});
