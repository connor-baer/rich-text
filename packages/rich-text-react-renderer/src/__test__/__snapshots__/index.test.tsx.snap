// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentToReactTree does not render unrecognized marks 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders asset hyperlink 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders blockquotes 1`] = `
Array [
  <p>
    [Function]
  </p>,
  <blockquote>
    [Function]
  </blockquote>,
]
`;

exports[`documentToReactTree renders default entry link block 1`] = `
Array [
  <div>
    [Function]
  </div>,
]
`;

exports[`documentToReactTree renders embedded entry 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders empty node if type is not recognized 1`] = `
Array [
  null,
]
`;

exports[`documentToReactTree renders entry hyperlink 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders horizontal rule 1`] = `
Array [
  <p>
    [Function]
  </p>,
  <hr />,
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders hyperlink 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders marks with default mark renderer 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders marks with default mark renderer 2`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders marks with default mark renderer 3`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders marks with default mark renderer 4`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders marks with the passed custom mark rendered 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders multiple marks with default mark renderer 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders nodes with default node renderer 1`] = `
Array [
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders nodes with default node renderer 2`] = `
Array [
  <h1>
    [Function]
  </h1>,
]
`;

exports[`documentToReactTree renders nodes with default node renderer 3`] = `
Array [
  <h2>
    [Function]
  </h2>,
]
`;

exports[`documentToReactTree renders nodes with passed custom node renderer 1`] = `
Array [
  <Paragraph>
    [Function]
  </Paragraph>,
]
`;

exports[`documentToReactTree renders ordered lists 1`] = `
Array [
  <ol>
    [Function]
  </ol>,
  <p>
    [Function]
  </p>,
]
`;

exports[`documentToReactTree renders unordered lists 1`] = `
Array [
  <ul>
    [Function]
  </ul>,
  <p>
    [Function]
  </p>,
]
`;
